catchup/peerSelector.go

```diff
@@ -0,0 +1,287 @@
+	if poolIdx < 0 || peerIdx < 0 {
+		return false
+	}
+
+	// we need to remove the peer from the pool so we can place it in a different location.
+	pool := ps.pools[poolIdx]
+	class := pool.peers[peerIdx].class
+	if len(pool.peers) > 1 {
+		pool.peers = append(pool.peers[:peerIdx], pool.peers[peerIdx+1:]...)
+		ps.pools[poolIdx] = pool
+	} else {
+		// the last peer was removed from the pool; delete this pool.
+		ps.pools = append(ps.pools[:poolIdx], ps.pools[poolIdx+1:]...)
+	}
+
+	sortNeeded := ps.addToPool(peer, rank, class)
+	if sortNeeded {
+		ps.sort()
+	}
+	return true
+}
+
+// PeerDownloadDurationToRank calculates the rank for a peer given a peer and the block download time.
+func (ps *peerSelector) PeerDownloadDurationToRank(peer network.Peer, blockDownloadDuration time.Duration) (rank int) {
+	ps.mu.Lock()
+	defer ps.mu.Unlock()
+	poolIdx, peerIdx := ps.findPeer(peer)
+	if poolIdx < 0 || peerIdx < 0 {
+		return peerRankInvalidDownload
+	}
+
+	switch ps.pools[poolIdx].peers[peerIdx].class.initialRank {
+	case peerRankInitialFirstPriority:
+		return downloadDurationToRank(blockDownloadDuration, lowBlockDownloadThreshold, highBlockDownloadThreshold, peerRank0LowBlockTime, peerRank0HighBlockTime)
+	default: // i.e. peerRankInitialSecondPriority
+		return downloadDurationToRank(blockDownloadDuration, lowBlockDownloadThreshold, highBlockDownloadThreshold, peerRank1LowBlockTime, peerRank1HighBlockTime)
+	}
+}
+
+// addToPool adds a given peer to the correct group. If no group exists for that peer's rank,
+// a new group is created.
+// The method return true if a new group was created ( suggesting that the pools list would need to be re-ordered ), or false otherwise.
+func (ps *peerSelector) addToPool(peer network.Peer, rank int, class peerClass) bool {
+	// see if we already have a list with that rank:
+	for i, pool := range ps.pools {
+		if pool.rank == rank {
+			// we found an existing group, add this peer to the list.
+			ps.pools[i] = peerPool{rank: rank, peers: append(pool.peers, peerPoolEntry{peer: peer, class: class})}
+			return false
+		}
+	}
+	ps.pools = append(ps.pools, peerPool{rank: rank, peers: []peerPoolEntry{{peer: peer, class: class}}})
+	return true
+}
+
+// sort the pools array in an accending order according to the rank of each pool.
+func (ps *peerSelector) sort() {
+	sort.SliceStable(ps.pools, func(i, j int) bool {
+		return ps.pools[i].rank < ps.pools[j].rank
+	})
+}
+
+// peerAddress returns the peer's underlying address. The network.Peer object cannot be compared
+// to itself, since the network package dynamically creating a new instance on every network.GetPeers() call.
+// The method retrun the peer address or an empty string if the peer is not one of HTTPPeer/UnicastPeer
+func peerAddress(peer network.Peer) string {
+	if httpPeer, ok := peer.(network.HTTPPeer); ok {
+		return httpPeer.GetAddress()
+	} else if unicastPeer, ok := peer.(network.UnicastPeer); ok {
+		return unicastPeer.GetAddress()
+	}
+	return ""
```

Comment:
what else can be in a peers list?

Reply:
There *shouldn't* be anything else. however, since the Peer is opaque interface, we need to verify. The robustness of this function allows us to avoid checking that in other places.